### PR TITLE
Refactor NQueens implementation

### PR DESCRIPTION
**Pull request by TruongNhanNguyen**
## Changes

- Rewrite `NQueensSolver` in OOP style
- Use `n_queens_solver` method to hide internal state of `NQueensSolver` object
- Add Rust docstrings
- Write parametrized tests, add 6x6 case

## Checklist:

- [x] I ran bellow commands using the latest version of **rust nightly**.
- [x] I ran `cargo clippy --all -- -D warnings` just before my last commit and fixed any issue that was found.
- [x] I ran `cargo fmt` just before my last commit.
- [x] I ran `cargo test` just before my last commit and all tests passed.
- [x] I added my algorithm to the corresponding `mod.rs` file within its own folder, and in any parent folder(s).
- [x] I added my algorithm to `DIRECTORY.md` with the correct link.
- [x] I checked `COUNTRIBUTING.md` and my code follows its guidelines.

### CODE REVIEW

**Comment by TruongNhanNguyen** [comment]
Can you have a look on this PR, @vil02?

**Comment by TruongNhanNguyen** [comment]
Can you review this PR, @vil02?

File: src/backtracking/n_queens.rs [hunk 1]
```diff
@@ -1,81 +1,108 @@
...
-            return false;
-        }
+    /// Solves the N-Queens problem and returns a vector of solutions.
+    ///
+    /// # Returns
+    ///
+    /// A vector containing all solutions to the N-Queens problem.
+    fn solve(&mut self) -> Vec<Vec<String>> {
+        self.solve_helper(0);
+        self.solutions.clone()
```
Comment (vil02) [inline src/backtracking/n_queens.rs:62]:
```suggestion
        std::mem::take(&mut self.solutions)
```

File: src/backtracking/n_queens.rs [hunk 2]
```diff
@@ -84,17 +111,32 @@ fn solve(board: &mut Vec<Vec<char>>, row: usize, solutions: &mut Vec<Vec<String>
...
+        ($($name:ident: $tc:expr,)*) => {
+            $(
+                #[test]
+                fn $name() {
+                    let (n, expected_solutions) = $tc;
+                    let solutions = n_queens_solver(n);
+                    assert_eq!(solutions.len(), expected_solutions.len());
+                    for (actual, expected) in solutions.iter().zip(expected_solutions.iter()) {
+                        assert_eq!(actual, expected);
+                    }
```
Comment (vil02) [inline src/backtracking/n_queens.rs:124]:
```suggestion
                    assert_eq!(solutions, expected_solutions);
```

File: src/backtracking/n_queens.rs [hunk 3]
```diff
@@ -84,17 +111,32 @@ fn solve(board: &mut Vec<Vec<char>>, row: usize, solutions: &mut Vec<Vec<String>
...
+        test_4_queens: (4, vec![
+            vec![".Q..", "...Q", "Q...", "..Q."],
+            vec!["..Q.", "Q...", "...Q", ".Q.."],
+        ]),
+        test_6_queens: (6, vec![
+            vec![".Q....", "...Q..", ".....Q", "Q.....", "..Q...", "....Q."],
+            vec!["..Q...", ".....Q", ".Q....", "....Q.", "Q.....", "...Q.."],
+            vec!["...Q..", "Q.....", "....Q.", ".Q....", ".....Q", "..Q..."],
+            vec!["....Q.", "..Q...", "Q.....", ".....Q", "...Q..", ".Q...."],
+        ]),
```
Comment (vil02) [inline src/backtracking/n_queens.rs:140]:
```suggestion
    test_n_queens_solver! {
        test_0_queens: (0, vec![Vec::<String>::new()]),
        test_1_queen: (1, vec![vec!["Q"]]),
        test_2_queens:(2, Vec::<Vec<String>>::new()),
        test_3_queens:(3, Vec::<Vec<String>>::new()),
        test_4_queens: (4, vec![
            vec![".Q..",
                 "...Q",
                 "Q...",
                 "..Q."],
            vec!["..Q.",
                 "Q...",
                 "...Q",
                 ".Q.."],
        ]),
        test_5_queens:(5, vec![
            vec!["Q....",
                 "..Q..",
                 "....Q",
                 ".Q...",
                 "...Q."],
            vec!["Q....",
                 "...Q.",
                 ".Q...",
                 "....Q",
                 "..Q.."],
            vec![".Q...",
                 "...Q.",
                 "Q....",
                 "..Q..",
                 "....Q"],
            vec![".Q...",
                 "....Q",
                 "..Q..",
                 "Q....",
                 "...Q."],
            vec!["..Q..",
                 "Q....",
                 "...Q.",
                 ".Q...",
                 "....Q"],
            vec!["..Q..",
                 "....Q",
                 ".Q...",
                 "...Q.",
                 "Q...."],
            vec!["...Q.",
                 "Q....",
                 "..Q..",
                 "....Q",
                 ".Q..."],
            vec!["...Q.",
                 ".Q...",
                 "....Q",
                 "..Q..",
                 "Q...."],
            vec!["....Q",
                 ".Q...",
                 "...Q.",
                 "Q....",
                 "..Q.."],
            vec!["....Q",
                 "..Q..",
                 "Q....",
                 "...Q.",
                 ".Q..."],
        ]),
        test_6_queens: (6, vec![
            vec![".Q....",
                 "...Q..",
                 ".....Q",
                 "Q.....",
                 "..Q...",
                 "....Q."],
            vec!["..Q...",
                 ".....Q",
                 ".Q....",
                 "....Q.",
                 "Q.....",
                 "...Q.."],
            vec!["...Q..",
                 "Q.....",
                 "....Q.",
                 ".Q....",
                 ".....Q",
                 "..Q..."],
            vec!["....Q.",
                 "..Q...",
                 "Q.....",
                 ".....Q",
                 "...Q..",
                 ".Q...."],
        ]),
```

**Review by vil02 — CHANGES_REQUESTED** [review]

File: src/backtracking/n_queens.rs (hunk 1 above)
Comment (TruongNhanNguyen) [reply src/backtracking/n_queens.rs:62]: It's better than `clone` the entire solution. Good point.

**Review by vil02 — APPROVED** [review]

**Comment by vil02** [comment]
>     * Rewrite `NQueensSolver` in OOP style

By the way: there is not much of OOP in this implementation (which is nothing bad). It is _class based_ at most.

**Comment by TruongNhanNguyen** [comment]
> > ```
> > * Rewrite `NQueensSolver` in OOP style
> > ```
> 
> By the way: there is not much of OOP in this implementation (which is nothing bad). It is _class based_ at most.

Thanks for pointing this out. I overused the OOP word in this context.

**Pull request merged**